### PR TITLE
[JUJU-2355] CI: Speed up go generate

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following attempt to speed up github/generate workflow. The go generate command takes forever to trawl the codebase, so instead, we'll use grep to select the go generates and do it manually.

We then have the option of parallelizing the generate commands.

## Checklist

- [x] Comments saying why design decisions were made

## QA steps

See the github action.

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]
Looks good, can we land in 3.0 though?

**Comment by barrettj12** [comment]
This is only giving 3 minutes speed up, I'm not sure if that's worth the extra complexity. If we can parallelise to speed it up further, then let's do that in this PR.

I'd also like to be smarter if possible about not regenerating files if we can be sure they haven't changed. We should only need to regenerate a subset of the files on most PRs (unless it's updating mockgen or something).

**Comment by hpidcock** [comment]
> This is only giving 3 minutes speed up, I'm not sure if that's worth the extra complexity. If we can parallelise to speed it up further, then let's do that in this PR.
> 
> I'd also like to be smarter if possible about not regenerating files if we can be sure they haven't changed. We should only need to regenerate a subset of the files on most PRs (unless it's updating mockgen or something).

`-P 2` on xargs should be all we need then.

**Comment by SimonRichardson** [comment]
@barrettj12 so I saw some runs being 40 minutes, hence why this PR exists.

> I'd also like to be smarter if possible about not regenerating files if we can be sure they haven't changed.

The problem you've got with this is that there are some generated mocks that live in different locations, that reference an interface outside of the package. Traversing for those changes seems like a lot of work for not much gain.

**Comment by barrettj12** [comment]
@SimonRichardson last run is 21 minutes, that's a great improvement.

> The problem you've got with this is that there are some generated mocks that live in different locations, that reference an interface outside of the package. Traversing for those changes seems like a lot of work for not much gain.

Yes, and there is a real lack of regularity about where the go generate directives are vs where the generated mocks end up. If we could agree on a pattern here, then I think it might be more feasible.

File: .github/workflows/gen.yml [hunk 1]
```diff
@@ -40,7 +40,8 @@ jobs:
 
       - name: Regenerate code
         shell: bash
-        run: go generate -x ./...
+        run: |
+          grep -ir "//go:generate" . | awk -F : '{ print $1 }' | uniq | xargs -n 1 -P 2 -I% go generate -x $(realpath %)
```
Comment (barrettj12) [inline .github/workflows/gen.yml:44]: This is pretty opaque, maybe a comment explaining what is going on here? Or break it up using variables?

**Review by barrettj12 — APPROVED** [review]

**Pull request merged**